opentracing/dapperish/main.go

```diff
@@ -0,0 +1,68 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"html"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"runtime"
+
+	"github.com/opentracing/api-golang/opentracing"
+)
+
+func client() {
+	reader := bufio.NewReader(os.Stdin)
+	for {
+		span, ctx := opentracing.StartSpan("getInput")
+		span.Info("ctx: ", ctx)
+		fmt.Print("Enter text: ")
+		text, _ := reader.ReadString('\n')
+		span.Info(text)
+
+		httpClient := &http.Client{}
+		httpReq, _ := http.NewRequest("POST", "http://localhost:8080/", bytes.NewReader([]byte(text)))
+		opentracing.AddContextIDToHttpHeader(span.ContextID(), httpReq.Header)
+		resp, err := httpClient.Do(httpReq)
+		if err != nil {
+			span.Error("error: ", err)
+		} else {
+			span.Info("got response: ", resp)
+		}
+
+		span.Finish()
+	}
+}
+
+func server() {
+	http.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
+		reqCtxID, err := opentracing.GetContextIDFromHttpHeader(
+			req.Header, opentracing.GlobalTracer())
+		if err != nil {
+			panic(err)
+		}
+
+		serverSpan, _ := opentracing.StartSpan("serverSpan", reqCtxID)
```

Comment:
Oh, I see. Yes, I would have just modeled these [naively, as you say] as fully nested spans. I had misunderstood previously and thought the client was doing some smart multi-hop stuff on its own.

`<soapbox>`
BTW, I have seen that paper and frankly have to roll my eyes, at least a little... *The single biggest problem* with tracing systems is rarely their data model: it is getting busy, overworked programmers to adopt them. The paper concedes that spans are nice because programmers have a pre-existing mental model for them. The paper then goes on to explain how the event model is more expressive, which – given the paper's definition of the span model – is obvious and hardly requires a formal proof!

But it's entirely possible to express multiple-parent DAG structures using the model we are discussing here... i.e., we can define a particular sort of `RawLog` that has a payload value for key `HappensBeforeSpanId`. All of a sudden, the span model can express anything the event model can express, and – voila – they are equivalent in terms of their expressive capabilities. Except that the default mode of operation in the span model is easier for busy programmers to understand at a glance.

As the paper puts it:

```The downside of the X-Trace model is that it is not necessarily as intuitive for programmers to use as the span model. The call-stack view that the span model provides has a level of simplicity that the event model cannot match.```

Right. So let's just use a span-based programming model which can provide for event-oriented features as a special case... thus "making the easy things easy and the hard things possible" (as they say).
`</soapbox>`